### ISSUE TITLE
LunarGOO producing incorrect result for shader and Unsupported functionality

### ISSUE THREAD

**Issue by danginsburg** [LunarG/LunarGLASS#26]
The following shader:

```
#version 140

uniform int g_nDebugMode;
uniform int g_nTest;
uniform float g_flTest;

uniform vec4 result0;
uniform vec4 result1;
uniform vec4 result2;
uniform vec4 result3;
vec4 ApplyDebugMode( out bool bReturn )
{
	vec4 vResult;
	if ( g_nDebugMode == 0 )
	{
		bReturn = true;
		vResult = result0;
	}
	else
	{
		bReturn = false;
		vResult = result1;
	}
	return vResult;
}

vec4 foo()
{
	vec4 vOutColor;
	if ( g_nDebugMode != 1 )
	{
		bool bReturn = true;
		vOutColor = ApplyDebugMode( bReturn );
		if ( bReturn )
			return vOutColor;
	}
	
	if ( g_flTest > 0.0 )
	{
		if ( g_nTest > 0 )
		{
			vOutColor = result2;
		}
		else
		{
			vOutColor = result3;
		}
			
	}
	return vOutColor;
}

void main()
{
	gl_FragColor = foo();
}
```

Produces the following output in LunarGOO:

***Unsupported functionality: general-case cross edges (requires artificial loop
)

And then also produces incorrect output code if the warning is ignored.

**Comment by danginsburg** [comment]
Also, rewriting the shader as follows works around the bug so it appears to be related to the early return after ApplyDebugMode.

```
#version 140

uniform int g_nDebugMode;
uniform int g_nTest;
uniform float g_flTest;

uniform vec4 result0;
uniform vec4 result1;
uniform vec4 result2;
uniform vec4 result3;
vec4 ApplyDebugMode( out bool bReturn )
{
	vec4 vResult;
	if ( g_nDebugMode == 0 )
	{
		bReturn = true;
		vResult = result0;
	}
	else
	{
		bReturn = false;
		vResult = result1;
	}
	return vResult;
}

vec4 foo()
{
	vec4 vOutColor;
	bool bReturn = false;
	if ( g_nDebugMode != 1 )
	{
		bReturn = true;
		vOutColor = ApplyDebugMode( bReturn );
		
	}
	
	if ( !bReturn && g_flTest > 0.0 )
	{
		if ( g_nTest > 0 )
		{
			vOutColor = result2;
		}
		else
		{
			vOutColor = result3;
		}
			
	}
	return vOutColor;
}

void main()
{
	gl_FragColor = foo();
}
```

**Comment by greg-lunarg** [comment]
Sorry its taken so long to get to this.

Running this shader on the LLVM3.6.2 branch does not abort and looks to generate correct code. The Top IR for both branches (master and LLVM3.6.2) is identical. So it seems to be a bug in LLVM 3.4, likely an issue with inlining a function with an early return.

So the choices are to try to fix the bug in 3.4 or change the Top IR for foo to mimic Dan's workaround and get rid of the early return instruction. I will probably try fixing 3.4 first.
